components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/AuthzUtil.java

```diff
@@ -4351,10 +4378,29 @@ private static String appendAuthenticatedIDPs(SessionDataCacheEntry sessionDataC
      * @param resultFromLogin The session context.
      * @param cookieValue The cookie string which contains the commonAuthId value.
      */
-    private static void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogin, String cookieValue) {
+    private static void associateAuthenticationHistory(SessionDataCacheEntry resultFromLogin, String cookieValue)
+            throws OAuthSystemException {
 
-        SessionContext sessionContext = getSessionContext(cookieValue,
-                resultFromLogin.getoAuth2Parameters().getLoginTenantDomain());
+        String tenantDomain = resultFromLogin.getoAuth2Parameters().getLoginTenantDomain();
+        /*
+         If the app is created in an organization, get the tenant domain of the primary organization since the
+         session is stored against the primary organization.
+        */
+        try {
+            String appTenantDomain = OAuth2Util.getTenantDomainOfOauthApp(
+                    resultFromLogin.getoAuth2Parameters().getClientId(), tenantDomain);
+            if (OrganizationManagementUtil.isOrganization(appTenantDomain)) {
+                String appOrgId = OAuth2ServiceComponentHolder.getInstance().getOrganizationManager()
+                        .resolveOrganizationId(appTenantDomain);
+                String primaryOrgId = OAuth2ServiceComponentHolder.getInstance().getOrganizationManager()
+                        .getPrimaryOrganizationId(appOrgId);
+                tenantDomain = OAuth2ServiceComponentHolder.getInstance().getOrganizationManager()
+                        .resolveTenantDomain(primaryOrgId);
+            }
+        } catch (OrganizationManagementException | IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new OAuthSystemException(e);
+        }
+        SessionContext sessionContext = getSessionContext(cookieValue, tenantDomain);
```

Comment:
Not related to this place. 
How will be session context will be stored? Will need to check the session retrieval, session extension